<a href="https://colab.research.google.com/github/kmendesDev/project_titanic/blob/Data_transformation_Cassio/data_cleaning_titanic.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Projeto Titanic

## Introdução

👋🛳️ Olá, seja bem-vindo ao Desafio Titanic Kaggle! Esta é a lendária competição de Machine Learning do Titanic - o melhor e primeiro desafio de ML no Kaggle.

Esse projeto foi desenvolvido em colaboração com os Cientistas de Dados:

:iphone: Developers contacts

Kaio Mendes - [![General badge](https://img.shields.io/badge/LinkedIn-0077B5?style=for-the-badge&logo=linkedin&logoColor=white<SUBJECT>-<STATUS>-<COLOR>.svg)](https://www.linkedin.com/in/kaio-mendes-bba196150)  -  [![General badge](https://img.shields.io/badge/GitHub-100000?style=for-the-badge&logo=github&logoColor=white<SUBJECT>-<STATUS>-<COLOR>.svg)](https://github.com/kmendesDev)

Cassio Viana - [![General badge](https://img.shields.io/badge/LinkedIn-0077B5?style=for-the-badge&logo=linkedin&logoColor=white<SUBJECT>-<STATUS>-<COLOR>.svg)](https://www.linkedin.com/in/cvs1987)  -  [![General badge](https://img.shields.io/badge/GitHub-100000?style=for-the-badge&logo=github&logoColor=white<SUBJECT>-<STATUS>-<COLOR>.svg)](https://github.com/cvs2010)


O objetivo desse projeto é utilizar aprendizado de máquina para criar um modelo que preveja quais passageiros sobreviveram ao naufrágio do Titanic.

O Desafio: O naufrágio do Titanic é um dos naufrágios mais infames da história. Em 15 de abril de 1912, durante sua viagem inaugural, o RMS Titanic, amplamente considerado "inafundável", afundou após colidir com um iceberg. Infelizmente, não havia botes salva-vidas suficientes para todos a bordo, resultando na morte de 1502 dos 2224 passageiros e tripulantes.

Embora tenha havido algum elemento de sorte envolvido na sobrevivência, parece que alguns grupos de pessoas tinham mais probabilidade de sobreviver do que outros.

Neste projeto, iremos desenvolver um modelo preditivo de Machine Learning que responda à pergunta: "que tipo de pessoas tinham mais probabilidade de sobreviver?" usando dados dos passageiros (como nome, idade, gênero, classe socioeconômica, etc).

Quais Dados foram usados?
Dois conjuntos de dados semelhantes que incluem informações dos passageiros, como nome, idade, gênero, classe socioeconômica, etc. Um conjunto de dados é intitulado train.csv e o outro é intitulado test.csv.

O train.csv contém os detalhes de um subconjunto dos passageiros a bordo (exatamente 891) e, o que é mais importante, revelará se eles sobreviveram ou não, também conhecido como "verdade básica". Já o conjunto de dados test.csv contém informações semelhantes, mas não revela a "verdade básica" para cada passageiro. Sua tarefa é prever esses resultados.

Usando os padrões encontrados nos dados do train.csv, prevemos se os 418 passageiros a bordo (encontrados no test.csv) sobreviveram.

Confira a pasta "data" para explorar ainda mais os conjuntos de dados.
https://www.kaggle.com/competitions/titanic/overview/description


##Importação de bibliotecas

In [264]:
import pandas as pd # Para manusear dados tabulares
import seaborn as srn # Para gráficos estatísticos
import statistics as sts # Para cálculos matemáticos
import numpy as np # Para operações numéricas
import seaborn as sns # Para visualização de dados estatísticos
import matplotlib.pyplot as plt # Para criação de gráficos

##Importação dos dados

In [265]:
# Baixar arquivos salvos no github
!wget https://github.com/kmendesDev/project_titanic/raw/main/data/train.csv
!wget https://github.com/kmendesDev/project_titanic/raw/main/data/test.csv
dataset_test = pd.read_csv("/content/test.csv")
dataset_train = pd.read_csv("/content/train.csv")

--2023-08-21 13:09:07--  https://github.com/kmendesDev/project_titanic/raw/main/data/train.csv
Resolving github.com (github.com)... 140.82.113.3
Connecting to github.com (github.com)|140.82.113.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/kmendesDev/project_titanic/main/data/train.csv [following]
--2023-08-21 13:09:07--  https://raw.githubusercontent.com/kmendesDev/project_titanic/main/data/train.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 61194 (60K) [text/plain]
Saving to: ‘train.csv.2’

train.csv.2         100%[===================>]  59.76K  --.-KB/s    in 0.009s  

2023-08-21 13:09:07 (6.84 MB/s) - ‘train.csv.2’ saved [61194/61194]

--2023-08-21 13:09:07--  https://github.com/k

In [266]:
dataset_train

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [267]:
dataset_train.tail()

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

In [268]:
dataset_train.shape

(891, 12)

In [269]:
dataset_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [270]:
dataset_train.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [271]:
dataset_test

PassengerId  Pclass                                          Name  \
0            892       3                              Kelly, Mr. James   
1            893       3              Wilkes, Mrs. James (Ellen Needs)   
2            894       2                     Myles, Mr. Thomas Francis   
3            895       3                              Wirz, Mr. Albert   
4            896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)   
..           ...     ...                                           ...   
413         1305       3                            Spector, Mr. Woolf   
414         1306       1                  Oliva y Ocana, Dona. Fermina   
415         1307       3                  Saether, Mr. Simon Sivertsen   
416         1308       3                           Ware, Mr. Frederick   
417         1309       3                      Peter, Master. Michael J   

        Sex   Age  SibSp  Parch              Ticket      Fare Cabin Embarked  
0      male  34.5      0      0              330911    7.8292   NaN        Q  
1    female  47.0      1      0              363272    7.0000   NaN        S  
2      male  62.0      0      0              240276    9.6875   NaN        Q  
3      male  27.0      0      0              315154    8.6625   NaN        S  
4    female  22.0      1      1             3101298   12.2875   NaN        S  
..      ...   ...    ...    ...                 ...       ...   ...      ...  
413    male   NaN      0      0           A.5. 3236    8.0500   NaN        S  
414  female  39.0      0      0            PC 17758  108.9000  C105        C  
415    male  38.5      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416    male   NaN      0      0              359309    8.0500   NaN        S  
417    male   NaN      1      1                2668   22.3583   NaN        C  

[418 rows x 11 columns]

In [272]:
dataset_test.tail()

PassengerId  Pclass                          Name     Sex   Age  SibSp  \
413         1305       3            Spector, Mr. Woolf    male   NaN      0   
414         1306       1  Oliva y Ocana, Dona. Fermina  female  39.0      0   
415         1307       3  Saether, Mr. Simon Sivertsen    male  38.5      0   
416         1308       3           Ware, Mr. Frederick    male   NaN      0   
417         1309       3      Peter, Master. Michael J    male   NaN      1   

     Parch              Ticket      Fare Cabin Embarked  
413      0           A.5. 3236    8.0500   NaN        S  
414      0            PC 17758  108.9000  C105        C  
415      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416      0              359309    8.0500   NaN        S  
417      1                2668   22.3583   NaN        C

In [273]:
dataset_test.shape

(418, 11)

In [274]:
dataset_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Pclass       418 non-null    int64  
 2   Name         418 non-null    object 
 3   Sex          418 non-null    object 
 4   Age          332 non-null    float64
 5   SibSp        418 non-null    int64  
 6   Parch        418 non-null    int64  
 7   Ticket       418 non-null    object 
 8   Fare         417 non-null    float64
 9   Cabin        91 non-null     object 
 10  Embarked     418 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 36.0+ KB


In [275]:
dataset_test.describe()

PassengerId      Pclass         Age       SibSp       Parch        Fare
count   418.000000  418.000000  332.000000  418.000000  418.000000  417.000000
mean   1100.500000    2.265550   30.272590    0.447368    0.392344   35.627188
std     120.810458    0.841838   14.181209    0.896760    0.981429   55.907576
min     892.000000    1.000000    0.170000    0.000000    0.000000    0.000000
25%     996.250000    1.000000   21.000000    0.000000    0.000000    7.895800
50%    1100.500000    3.000000   27.000000    0.000000    0.000000   14.454200
75%    1204.750000    3.000000   39.000000    1.000000    0.000000   31.500000
max    1309.000000    3.000000   76.000000    8.000000    9.000000  512.329200

##Análise exploratória dos dados

### Análise do conjunto de dados de treinamento

In [276]:
#Clonando o DataFrame
training_data = dataset_train
training_data

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [277]:
# Informa a identificação do passageiro na lista
training_data['PassengerId'].describe()

count    891.000000
mean     446.000000
std      257.353842
min        1.000000
25%      223.500000
50%      446.000000
75%      668.500000
max      891.000000
Name: PassengerId, dtype: float64

In [278]:
# Coluna Survived:
# Informa a quantidade de 0 e 1 indicando com 1 os que sobreviveram e com 0 os que não conseguiram.
training_data.groupby(['Survived']).size()

Survived
0    549
1    342
dtype: int64

In [279]:
# Verificando a quantidade de itens na coluna Survived
training_data['Survived'].describe()

count    891.000000
mean       0.383838
std        0.486592
min        0.000000
25%        0.000000
50%        0.000000
75%        1.000000
max        1.000000
Name: Survived, dtype: float64

In [280]:
# Coluna Pclass:
# Informa a classe na qual o passageiro viajava:
training_data.groupby(['Pclass']).size()

Pclass
1    216
2    184
3    491
dtype: int64

In [281]:
# Conferindo a quantidade.
training_data['Pclass'].describe()

count    891.000000
mean       2.308642
std        0.836071
min        1.000000
25%        2.000000
50%        3.000000
75%        3.000000
max        3.000000
Name: Pclass, dtype: float64

In [282]:
# Conferindo a quantidade.
training_data['Name'].describe()

count                         891
unique                        891
top       Braund, Mr. Owen Harris
freq                            1
Name: Name, dtype: object

In [283]:
# Coluna sex:
# Informa a quantidade de homens e mulheres:
training_data['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

In [284]:
# Conferindo a quantidade.
training_data['Sex'].describe()

count      891
unique       2
top       male
freq       577
Name: Sex, dtype: object

In [285]:
# Coluna Age:
# Verificar a idade dos passageiros:
training_data['Age'].describe()
# A quantidade está inferior à 891. Logo temos problemas com quantitativo nessa coluna

count    714.000000
mean      29.699118
std       14.526497
min        0.420000
25%       20.125000
50%       28.000000
75%       38.000000
max       80.000000
Name: Age, dtype: float64

In [286]:
# Coluna SibSp:
# Informa a quantidade de irmãos e/ou cônjuges à bordo:
training_data['SibSp'].value_counts()

0    608
1    209
2     28
4     18
3     16
8      7
5      5
Name: SibSp, dtype: int64

In [287]:
# Conferindo a quantidade.
training_data['SibSp'].describe()

count    891.000000
mean       0.523008
std        1.102743
min        0.000000
25%        0.000000
50%        0.000000
75%        1.000000
max        8.000000
Name: SibSp, dtype: float64

In [288]:
# Coluna Parch:
# Informa a quantidade de pais e/ou filhos à bordo:
training_data['Parch'].value_counts()

0    678
1    118
2     80
5      5
3      5
4      4
6      1
Name: Parch, dtype: int64

In [289]:
# Conferindo a quantidade.
training_data['Parch'].describe()

count    891.000000
mean       0.381594
std        0.806057
min        0.000000
25%        0.000000
50%        0.000000
75%        0.000000
max        6.000000
Name: Parch, dtype: float64

In [290]:
# Coluna Ticket:
# Informa a numeração do Ticket de cada indivíduo:
training_data['Ticket'].value_counts()

347082      7
CA. 2343    7
1601        7
3101295     6
CA 2144     6
           ..
9234        1
19988       1
2693        1
PC 17612    1
370376      1
Name: Ticket, Length: 681, dtype: int64

In [291]:
# Conferindo a quantidade.
training_data['Ticket'].describe()

count        891
unique       681
top       347082
freq           7
Name: Ticket, dtype: object

In [292]:
# Coluna Fare:
# Informa a tarifa paga:
training_data['Fare'].describe()

count    891.000000
mean      32.204208
std       49.693429
min        0.000000
25%        7.910400
50%       14.454200
75%       31.000000
max      512.329200
Name: Fare, dtype: float64

In [293]:
# Coluna Cabin:
# Informa a cabina que o passageiro ocupava:
training_data['Cabin'].value_counts()
# Verifica-se que a cabine mais recorrente é a B96

B96 B98        4
G6             4
C23 C25 C27    4
C22 C26        3
F33            3
              ..
E34            1
C7             1
C54            1
E36            1
C148           1
Name: Cabin, Length: 147, dtype: int64

In [294]:
# Conferindo a quantidade.
training_data['Cabin'].describe()
# Verifica-se que faltam muitos valores. Temos apenas 204 preenchidos

count         204
unique        147
top       B96 B98
freq            4
Name: Cabin, dtype: object

In [295]:
# Coluna Embarked:
# Informa o porto o qual o passageiro embarcou:
training_data['Embarked'].value_counts()
# O Porto de Southampton (S) é o mais recorrente.

S    644
C    168
Q     77
Name: Embarked, dtype: int64

In [296]:
# Conferindo a quantidade.
training_data['Embarked'].describe()

count     889
unique      3
top         S
freq      644
Name: Embarked, dtype: object

Verificação de valores NaN

In [297]:
training_data.isnull().sum()
# Colunas com valores nulos:
# Age
# Cabin
# Embarked

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

Retirando os valores nulos:

In [298]:
# Coluna Age:
# Achando a mediana para substituir nos valores NaN:
mediana = sts.median(training_data['Age'])
mediana
# Problema: A mediana é NaN

nan

In [299]:
# Calculando a mediana sem considerar os valores NaN
mediana_sem_nan = np.nanmedian(training_data['Age'])

# Agora você tem a mediana sem os valores NaN
print(mediana_sem_nan)

28.0


In [300]:
#Substituindo os valores NaN pela média:
training_data['Age'].fillna(int(training_data['Age'].describe()[1]),inplace=True)
int(training_data['Age'].describe()[1])

29

In [301]:
#Verificando a existência de algum valor NaN:
training_data['Age'].isnull().sum()

0

In [302]:
# Verificando a mediana após o ajuste:
mediana = sts.median(training_data['Age'])
mediana

29.0

In [303]:
# Coluna Cabin:
training_data['Cabin'].isnull().sum()
#Verifica-se 687 valores nulos, caracterizando mais da metade.

687

In [304]:
# Coluna Embarked:
training_data['Embarked'].isnull().sum()
#Verifica-se 2 valores nulos na coluna Embarked.

2

In [305]:
training_data['Embarked'].fillna('S',inplace=True)

In [306]:
# Verificando se ainda há valores nulos:
training_data['Embarked'].isnull().sum()

0

In [307]:
#Verificando a nova quantidade de dados por variável:
training_data['Embarked'].value_counts()

S    646
C    168
Q     77
Name: Embarked, dtype: int64

Verificando valores duplicados de algumas colunas: Id, Name e Cabin

In [308]:
# Coluna PassengerId
training_data[training_data.duplicated(['PassengerId'],keep=False)]
# Verifica-se que não há valores duplicados

Empty DataFrame
Columns: [PassengerId, Survived, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, Embarked]
Index: []

In [309]:
# Coluna Name
training_data[training_data.duplicated(['Name'],keep=False)]
# Verifica-se que não há valores duplicados

Empty DataFrame
Columns: [PassengerId, Survived, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, Embarked]
Index: []

In [310]:
# Filtra os dados duplicados e não nulos na coluna 'Cabin'
filtered_data_cabin = training_data[training_data['Cabin'].duplicated(keep=False) & training_data['Cabin'].notna()]
# Ordena os dados
filtered_data_cabin = filtered_data_cabin.sort_values(by='Cabin')
# Excluindo o indice
filtered_data_cabin.reset_index(drop=True)
# Exibe os dados
filtered_data_cabin

PassengerId  Survived  Pclass  \
329          330         1       1   
523          524         1       1   
781          782         1       1   
690          691         1       1   
745          746         0       1   
..           ...       ...     ...   
618          619         1       2   
251          252         0       3   
10            11         1       3   
205          206         0       3   
394          395         1       3   

                                                  Name     Sex   Age  SibSp  \
329                       Hippach, Miss. Jean Gertrude  female  16.0      0   
523    Hippach, Mrs. Louis Albert (Ida Sophia Fischer)  female  44.0      0   
781          Dick, Mrs. Albert Adrian (Vera Gillespie)  female  17.0      1   
690                            Dick, Mr. Albert Adrian    male  31.0      1   
745                       Crosby, Capt. Edward Gifford    male  70.0      1   
..                                                 ...     ...   ...    ...   
618                        Becker, Miss. Marion Louise  female   4.0      2   
251         Strom, Mrs. Wilhelm (Elna Matilda Persson)  female  29.0      1   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
205                         Strom, Miss. Telma Matilda  female   2.0      0   
394  Sandstrom, Mrs. Hjalmar (Agnes Charlotta Bengt...  female  24.0      0   

     Parch     Ticket     Fare Cabin Embarked  
329      1     111361  57.9792   B18        C  
523      1     111361  57.9792   B18        C  
781      0      17474  57.0000   B20        S  
690      0      17474  57.0000   B20        S  
745      1  WE/P 5735  71.0000   B22        S  
..     ...        ...      ...   ...      ...  
618      1     230136  39.0000    F4        S  
251      1     347054  10.4625    G6        S  
10       1    PP 9549  16.7000    G6        S  
205      1     347054  10.4625    G6        S  
394      2    PP 9549  16.7000    G6        S  

[103 rows x 12 columns]

Alguns passageiros não tem cabine registradas: 687

Alguns passageiros tem cabines duplicadas: 103

Alguns passageiros tem mais de uma cabine registrada, provavelmente cabines conexas: 16

Passageiros registrados na mesma cabine tiveram diferentes taxas de sobrevivência, isso pode evidenciar algo relacionado ao posicionamento da pessoa no momento do naufrágio, ou outra característica como idade, dificuldade de locomoção, etc.

### Análise do conjunto de dados de teste

In [311]:
#Clonando o DataFrame teste
testing_data = dataset_test
testing_data

PassengerId  Pclass                                          Name  \
0            892       3                              Kelly, Mr. James   
1            893       3              Wilkes, Mrs. James (Ellen Needs)   
2            894       2                     Myles, Mr. Thomas Francis   
3            895       3                              Wirz, Mr. Albert   
4            896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)   
..           ...     ...                                           ...   
413         1305       3                            Spector, Mr. Woolf   
414         1306       1                  Oliva y Ocana, Dona. Fermina   
415         1307       3                  Saether, Mr. Simon Sivertsen   
416         1308       3                           Ware, Mr. Frederick   
417         1309       3                      Peter, Master. Michael J   

        Sex   Age  SibSp  Parch              Ticket      Fare Cabin Embarked  
0      male  34.5      0      0              330911    7.8292   NaN        Q  
1    female  47.0      1      0              363272    7.0000   NaN        S  
2      male  62.0      0      0              240276    9.6875   NaN        Q  
3      male  27.0      0      0              315154    8.6625   NaN        S  
4    female  22.0      1      1             3101298   12.2875   NaN        S  
..      ...   ...    ...    ...                 ...       ...   ...      ...  
413    male   NaN      0      0           A.5. 3236    8.0500   NaN        S  
414  female  39.0      0      0            PC 17758  108.9000  C105        C  
415    male  38.5      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416    male   NaN      0      0              359309    8.0500   NaN        S  
417    male   NaN      1      1                2668   22.3583   NaN        C  

[418 rows x 11 columns]

In [312]:
# Informa a identificação do passageiro na lista
testing_data['PassengerId'].describe()

count     418.000000
mean     1100.500000
std       120.810458
min       892.000000
25%       996.250000
50%      1100.500000
75%      1204.750000
max      1309.000000
Name: PassengerId, dtype: float64

In [313]:
# Coluna Pclass:
# Informa a classe na qual o passageiro viajava:
testing_data.groupby(['Pclass']).size()

Pclass
1    107
2     93
3    218
dtype: int64

In [314]:
# Conferindo a quantidade.
testing_data['Pclass'].describe()

count    418.000000
mean       2.265550
std        0.841838
min        1.000000
25%        1.000000
50%        3.000000
75%        3.000000
max        3.000000
Name: Pclass, dtype: float64

In [315]:
# Conferindo a quantidade.
testing_data['Name'].describe()

count                  418
unique                 418
top       Kelly, Mr. James
freq                     1
Name: Name, dtype: object

In [316]:
# Coluna sex:
# Informa a quantidade de homens e mulheres:
testing_data['Sex'].value_counts()

male      266
female    152
Name: Sex, dtype: int64

In [317]:
# Conferindo a quantidade.
testing_data['Sex'].describe()

count      418
unique       2
top       male
freq       266
Name: Sex, dtype: object

In [318]:
# Coluna Age:
# Verificar a idade dos passageiros:
testing_data['Age'].describe()
# A quantidade está inferior à 418. Logo temos problemas com quantitativo nessa coluna

count    332.000000
mean      30.272590
std       14.181209
min        0.170000
25%       21.000000
50%       27.000000
75%       39.000000
max       76.000000
Name: Age, dtype: float64

In [319]:
# Coluna SibSp:
# Informa a quantidade de irmãos e/ou cônjuges à bordo:
testing_data['SibSp'].value_counts()

0    283
1    110
2     14
3      4
4      4
8      2
5      1
Name: SibSp, dtype: int64

In [320]:
# Conferindo a quantidade.
testing_data['SibSp'].describe()

count    418.000000
mean       0.447368
std        0.896760
min        0.000000
25%        0.000000
50%        0.000000
75%        1.000000
max        8.000000
Name: SibSp, dtype: float64

In [321]:
# Coluna Parch:
# Informa a quantidade de pais e/ou filhos à bordo:
testing_data['Parch'].value_counts()

0    324
1     52
2     33
3      3
4      2
9      2
6      1
5      1
Name: Parch, dtype: int64

In [322]:
# Conferindo a quantidade.
testing_data['Parch'].describe()

count    418.000000
mean       0.392344
std        0.981429
min        0.000000
25%        0.000000
50%        0.000000
75%        0.000000
max        9.000000
Name: Parch, dtype: float64

In [323]:
# Coluna Ticket:
# Informa a numeração do Ticket de cada indivíduo:
testing_data['Ticket'].value_counts()

PC 17608    5
CA. 2343    4
113503      4
PC 17483    3
220845      3
           ..
349226      1
2621        1
4133        1
113780      1
2668        1
Name: Ticket, Length: 363, dtype: int64

In [324]:
# Conferindo a quantidade.
testing_data['Ticket'].describe()

count          418
unique         363
top       PC 17608
freq             5
Name: Ticket, dtype: object

In [325]:
# Coluna Fare:
# Informa a tarifa paga:
testing_data['Fare'].describe()

count    417.000000
mean      35.627188
std       55.907576
min        0.000000
25%        7.895800
50%       14.454200
75%       31.500000
max      512.329200
Name: Fare, dtype: float64

In [348]:
# Coluna Fare:
# Informa a tarifa paga:
testing_data['Fare'].isnull().sum()
# Apenas 1 valor nulo

1

In [326]:
# Coluna Cabin:
# Informa a cabina que o passageiro ocupava:
testing_data['Cabin'].value_counts()
# Verifica-se que a cabine mais recorrente é a B96

B57 B59 B63 B66    3
B45                2
C89                2
C55 C57            2
A34                2
                  ..
E52                1
D30                1
E31                1
C62 C64            1
C105               1
Name: Cabin, Length: 76, dtype: int64

In [327]:
# Conferindo a quantidade.
testing_data['Cabin'].describe()
# Verifica-se que faltam muitos valores. Temos apenas 204 preenchidos

count                  91
unique                 76
top       B57 B59 B63 B66
freq                    3
Name: Cabin, dtype: object

In [328]:
# Coluna Embarked:
# Informa o porto o qual o passageiro embarcou:
testing_data['Embarked'].value_counts()
# O Porto de Southampton (S) é o mais recorrente.

S    270
C    102
Q     46
Name: Embarked, dtype: int64

In [329]:
# Conferindo a quantidade.
testing_data['Embarked'].describe()

count     418
unique      3
top         S
freq      270
Name: Embarked, dtype: object

Verificação de valores NaN

In [330]:
testing_data.isnull().sum()
# Colunas com valores nulos:
# Age
# Cabin
# Embarked

PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64

Retirando os valores nulos:

In [331]:
# Coluna Age:
# Achando a mediana para substituir nos valores NaN:
mediana = sts.median(testing_data['Age'])
mediana
# Problema: A mediana é NaN

nan

In [332]:
# Calculando a mediana sem considerar os valores NaN
mediana_sem_nan = np.nanmedian(testing_data['Age'])

# Agora você tem a mediana sem os valores NaN
print(mediana_sem_nan)

27.0


In [333]:
#Substituindo os valores NaN pela média:
testing_data['Age'].fillna(int(testing_data['Age'].describe()[1]),inplace=True)
int(testing_data['Age'].describe()[1])

30

In [334]:
#Verificando a existência de algum valor NaN:
testing_data['Age'].isnull().sum()

0

In [335]:
# Verificando a mediana após o ajuste:
mediana = sts.median(testing_data['Age'])
mediana

30.0

In [336]:
# Coluna Cabin:
testing_data['Cabin'].isnull().sum()
#Verifica-se 327 valores nulos, caracterizando mais da metade.

327

In [337]:
# Coluna Embarked:
testing_data['Embarked'].isnull().sum()

0

Verificando valores duplicados de algumas colunas: Id, Name e Cabin

In [341]:
# Coluna PassengerId
testing_data[testing_data.duplicated(['PassengerId'],keep=False)]
# Verifica-se que não há valores duplicados

Empty DataFrame
Columns: [PassengerId, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, Embarked]
Index: []

In [342]:
# Coluna Name
testing_data[testing_data.duplicated(['Name'],keep=False)]
# Verifica-se que não há valores duplicados

Empty DataFrame
Columns: [PassengerId, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, Embarked]
Index: []

In [351]:
# Filtra os dados duplicados e não nulos na coluna 'Cabin'
filtered_data_cabin2 = testing_data[testing_data['Cabin'].duplicated(keep=False) & testing_data['Cabin'].notna()]
# Ordena os dados
filtered_data_cabin2 = filtered_data_cabin2.sort_values(by='Cabin')
# Excluindo o indice
filtered_data_cabin2.reset_index(drop=True)
# Exibe os dados
filtered_data_cabin2

PassengerId  Pclass                                               Name  \
293         1185       1                              Dodge, Dr. Washington   
374         1266       1              Dodge, Mrs. Washington (Ruth Vidaver)   
12           904       1      Snyder, Mrs. John Pillsbury (Nelle Stevenson)   
287         1179       1                         Snyder, Mr. John Pillsbury   
24           916       1    Ryerson, Mrs. Arthur Larned (Emily Maria Borie)   
64           956       1                        Ryerson, Master. John Borie   
142         1034       1                         Ryerson, Mr. Arthur Larned   
356         1248       1     Brown, Mrs. John Murray (Caroline Lane Lamson)   
77           969       1  Cornell, Mrs. Robert Clifford (Malvina Helen L...   
177         1069       1                    Stengel, Mr. Charles Emil Henry   
100          992       1  Stengel, Mrs. Charles Emil Henry (Annie May Mo...   
69           961       1                Fortune, Mrs. Mark (Mary McDougald)   
53           945       1                         Fortune, Miss. Ethel Flora   
50           942       1                           Smith, Mr. Lucien Philip   
395         1287       1     Smith, Mrs. Lucien Philip (Mary Eloise Hughes)   
81           973       1                                 Straus, Mr. Isidor   
114         1006       1             Straus, Mrs. Isidor (Rosalie Ida Blun)   
118         1010       1                               Beattie, Mr. Thomson   
270         1162       1                       McCaffry, Mr. Thomas Francis   
411         1303       1    Minahan, Mrs. William Edward (Lillian E Thorpe)   
34           926       1                           Mock, Mr. Philipp Edmund   
218         1110       1       Widener, Mrs. George Dunton (Eleanor Elkins)   
407         1299       1                         Widener, Mr. George Dunton   
272         1164       1      Clark, Mrs. Walter Miller (Virginia McDowell)   
252         1144       1                           Clark, Mr. Walter Miller   
242         1134       1                       Spedden, Mr. Frederic Oakley   
196         1088       1                    Spedden, Master. Robert Douglas   
326         1218       2                       Becker, Miss. Ruth Elizabeth   
178         1070       2   Becker, Mrs. Allen Oliver (Nellie E Baumgardner)   

        Sex   Age  SibSp  Parch    Ticket      Fare            Cabin Embarked  
293    male  53.0      1      1     33638   81.8583              A34        S  
374  female  54.0      1      1     33638   81.8583              A34        S  
12   female  23.0      1      0     21228   82.2667              B45        S  
287    male  24.0      1      0     21228   82.2667              B45        S  
24   female  48.0      1      3  PC 17608  262.3750  B57 B59 B63 B66        C  
64     male  13.0      2      2  PC 17608  262.3750  B57 B59 B63 B66        C  
142    male  61.0      1      3  PC 17608  262.3750  B57 B59 B63 B66        C  
356  female  59.0      2      0     11769   51.4792             C101        S  
77   female  55.0      2      0     11770   25.7000             C101        S  
177    male  54.0      1      0     11778   55.4417             C116        C  
100  female  43.0      1      0     11778   55.4417             C116        C  
69   female  60.0      1      4     19950  263.0000      C23 C25 C27        S  
53   female  28.0      3      2     19950  263.0000      C23 C25 C27        S  
50     male  24.0      1      0     13695   60.0000              C31        S  
395  female  18.0      1      0     13695   60.0000              C31        S  
81     male  67.0      1      0  PC 17483  221.7792          C55 C57        S  
114  female  63.0      1      0  PC 17483  221.7792          C55 C57        S  
118    male  36.0      0      0     13050   75.2417               C6        C  
270    male  46.0      0      0     13050   75.2417               C6        C  
411  female  37.0      1      0   

In [349]:
testing_data['Cabin'].isnull().sum()

327

In [352]:
filtered_data_cabin2.shape

(29, 11)

Alguns passageiros não tem cabine registradas: 327

Alguns passageiros tem cabines duplicadas: 29

Alguns passageiros tem mais de uma cabine registrada, provavelmente cabines conexas: 7

Passageiros registrados na mesma cabine tiveram diferentes taxas de sobrevivência, isso pode evidenciar algo relacionado ao posicionamento da pessoa no momento do naufrágio, ou outra característica como idade, dificuldade de locomoção, etc.

## Pré-processamento, limpeza e tratamento dos dados

### Unindo os datasets de treinamento e teste

É crucial lidar com desafios como substituição de valores em colunas com valores ausentes, como a coluna de idade com valores NaN. Isso frequentemente envolve calcular a média geral dos dados para realizar as substituições de forma adequada. Além disso, quando lidamos com colunas não numéricas, é essencial realizar o processo de labeling para transformar esses dados em formatos que os algoritmos possam compreender. Esse tipo de pré-processamento é fundamental para garantir a integridade dos dados e prepará-los para análises mais aprofundadas.

Como um cientista de dados, desenvolver insights como esses e aplicá-los durante a resolução de problemas e ajustes em projetos é uma habilidade essencial. Isso não apenas contribui para a qualidade das análises, mas também permite tomar decisões embasadas em dados confiáveis. Portanto, ter um conhecimento sólido sobre manipulação e tratamento de dados é uma base fundamental para o sucesso na área de ciência de dados.

Agora vamos criar novamente os datasets de treino e teste sem alterações e unir os dois com a finalidade de fazer o pré-processamento.

In [554]:
# Criando nova cópia dos datasets sem alteração
dataset_test = pd.read_csv("/content/test.csv")
dataset_train = pd.read_csv("/content/train.csv")
training_data = dataset_train
testing_data = dataset_test

# Unindo os datasets de treino e teste
# Concatenando os datasets verticalmente
dataset = pd.concat([training_data, testing_data], ignore_index=True)

# Exibindo os dados
dataset

PassengerId  Survived  Pclass  \
0               1       0.0       3   
1               2       1.0       1   
2               3       1.0       3   
3               4       1.0       1   
4               5       0.0       3   
...           ...       ...     ...   
1304         1305       NaN       3   
1305         1306       NaN       1   
1306         1307       NaN       3   
1307         1308       NaN       3   
1308         1309       NaN       3   

                                                   Name     Sex   Age  SibSp  \
0                               Braund, Mr. Owen Harris    male  22.0      1   
1     Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                                Heikkinen, Miss. Laina  female  26.0      0   
3          Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                              Allen, Mr. William Henry    male  35.0      0   
...                                                 ...     ...   ...    ...   
1304                                 Spector, Mr. Woolf    male   NaN      0   
1305                       Oliva y Ocana, Dona. Fermina  female  39.0      0   
1306                       Saether, Mr. Simon Sivertsen    male  38.5      0   
1307                                Ware, Mr. Frederick    male   NaN      0   
1308                           Peter, Master. Michael J    male   NaN      1   

      Parch              Ticket      Fare Cabin Embarked  
0         0           A/5 21171    7.2500   NaN        S  
1         0            PC 17599   71.2833   C85        C  
2         0    STON/O2. 3101282    7.9250   NaN        S  
3         0              113803   53.1000  C123        S  
4         0              373450    8.0500   NaN        S  
...     ...                 ...       ...   ...      ...  
1304      0           A.5. 3236    8.0500   NaN        S  
1305      0            PC 17758  108.9000  C105        C  
1306      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
1307      0              359309    8.0500   NaN        S  
1308      1                2668   22.3583   NaN        C  

[1309 rows x 12 columns]

Percebemos que a concatenação funcionou perfeitamente, o número total de passageiros é 1309.

### Pré-processamento e limpeza dos dados

In [555]:
dataset.isnull().sum()

PassengerId       0
Survived        418
Pclass            0
Name              0
Sex               0
Age             263
SibSp             0
Parch             0
Ticket            0
Fare              1
Cabin          1014
Embarked          2
dtype: int64

Existem valores ausentes nas colunas Survived (418, que representam valores ausentes pois só temos os dados de sobreviventes de passageiros do dataset de treinamento) , Age , Fare , Cabin e Embarked.

Vamos proceder com o tratamento dos dados ausentes

Anteriormente o conjunto de treinamento apresentava média e mediana 29

e o conjunto de teste apresentava média e mediana 30

In [556]:
# Calculando a mediana sem considerar os valores NaN
mediana_sem_nan = np.nanmedian(dataset['Age'])

# Agora você tem a mediana sem os valores NaN
print(mediana_sem_nan)

# Média de idade dos passageiros do dataset
print(int(dataset['Age'].describe()[1]))

28.0
29


O conjunto completo dos dados agora apresenta nova média de idade de 29

E a nova mediana de 28, teremos que decidir como lidar com os valores ausentes

Inicialmente vamos substituir coma média

In [557]:
# Vamos substituir os valores NaN pela média:
dataset['Age'].fillna(int(dataset['Age'].describe()[1]),inplace=True)
#Verificando a existência de algum valor NaN:
dataset['Age'].isnull().sum()

0

In [558]:
# Substituindo o valor ausentes em Fare pela média:
dataset['Fare'].fillna(int(dataset['Fare'].describe()[1]),inplace=True)
#Verificando a existência de algum valor NaN:
dataset['Fare'].isnull().sum()

0

In [559]:
#Verificando a nova quantidade de dados por variável:
dataset['Embarked'].value_counts()

S    914
C    270
Q    123
Name: Embarked, dtype: int64

In [560]:
# A maior parte dos passageiros embarcou no porto S
# Substituindo o valor ausentes em Embarked por S:
dataset['Embarked'].fillna('S',inplace=True)
#Verificando a existência de algum valor NaN:
dataset['Embarked'].isnull().sum()

0

In [561]:
# Substituindo os valores NaN de Cabin por zero:
dataset['Cabin'].fillna(0,inplace=True)
#Verificando a existência de algum valor NaN:
dataset['Cabin'].isnull().sum()

0

In [562]:
dataset

PassengerId  Survived  Pclass  \
0               1       0.0       3   
1               2       1.0       1   
2               3       1.0       3   
3               4       1.0       1   
4               5       0.0       3   
...           ...       ...     ...   
1304         1305       NaN       3   
1305         1306       NaN       1   
1306         1307       NaN       3   
1307         1308       NaN       3   
1308         1309       NaN       3   

                                                   Name     Sex   Age  SibSp  \
0                               Braund, Mr. Owen Harris    male  22.0      1   
1     Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                                Heikkinen, Miss. Laina  female  26.0      0   
3          Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                              Allen, Mr. William Henry    male  35.0      0   
...                                                 ...     ...   ...    ...   
1304                                 Spector, Mr. Woolf    male  29.0      0   
1305                       Oliva y Ocana, Dona. Fermina  female  39.0      0   
1306                       Saether, Mr. Simon Sivertsen    male  38.5      0   
1307                                Ware, Mr. Frederick    male  29.0      0   
1308                           Peter, Master. Michael J    male  29.0      1   

      Parch              Ticket      Fare Cabin Embarked  
0         0           A/5 21171    7.2500     0        S  
1         0            PC 17599   71.2833   C85        C  
2         0    STON/O2. 3101282    7.9250     0        S  
3         0              113803   53.1000  C123        S  
4         0              373450    8.0500     0        S  
...     ...                 ...       ...   ...      ...  
1304      0           A.5. 3236    8.0500     0        S  
1305      0            PC 17758  108.9000  C105        C  
1306      0  SOTON/O.Q. 3101262    7.2500     0        S  
1307      0              359309    8.0500     0        S  
1308      1                2668   22.3583     0        C  

[1309 rows x 12 columns]

In [563]:
dataset.isnull().sum()

PassengerId      0
Survived       418
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin            0
Embarked         0
dtype: int64

Confirmamos que só há valores ausentes na coluna Survived, completamos a primeira etapa de limpeza dos dados.

### Tratamento dos dados


Agora iremos para a segunda etapa de modificação das colunas com valores não numéricos. Essa etapa é chamada de "codificação" e visa transformar os valores não numéricos em formatos que possam ser interpretados por algoritmos de análise e modelagem. Isso é essencial para garantir que os dados sejam adequados para análises estatísticas e de aprendizado de máquina, permitindo que os modelos compreendam as informações presentes nas colunas não numéricas. Existem várias técnicas de codificação, como a codificação one-hot e a codificação de rótulos, que podem ser aplicadas dependendo das características dos dados e dos objetivos da análise.

In [564]:
# Lista de colunas numéricas
colunas_numericas = dataset.select_dtypes(include=[np.number]).columns.tolist()

# Lista de colunas não numéricas
colunas_nao_numericas = dataset.select_dtypes(exclude=[np.number]).columns.tolist()

print("Colunas Numéricas:")
print(colunas_numericas)

print("\nColunas Não Numéricas:")
print(colunas_nao_numericas)

Colunas Numéricas:
['PassengerId', 'Survived', 'Pclass', 'Age', 'SibSp', 'Parch', 'Fare']

Colunas Não Numéricas:
['Name', 'Sex', 'Ticket', 'Cabin', 'Embarked']


A coluna nome não precisa de nenhum tipo de tratamento pois será desconsiderada para a análise do Machine Learning.

Vamos proceder com o tratamento das demais colunas.

In [565]:
# Realizando one-hot encoding na coluna "Sex"
one_hot_encoded = pd.get_dummies(dataset['Sex'], prefix='Sex')

# Adicionando as colunas codificadas ao DataFrame original
dataset = pd.concat([dataset, one_hot_encoded], axis=1)

In [566]:
# Realizando one-hot encoding na coluna "Embarked"
one_hot_encoded = pd.get_dummies(dataset['Embarked'], prefix='Embarked')

# Adicionando as colunas codificadas ao DataFrame original
dataset = pd.concat([dataset, one_hot_encoded], axis=1)

In [567]:
dataset

PassengerId  Survived  Pclass  \
0               1       0.0       3   
1               2       1.0       1   
2               3       1.0       3   
3               4       1.0       1   
4               5       0.0       3   
...           ...       ...     ...   
1304         1305       NaN       3   
1305         1306       NaN       1   
1306         1307       NaN       3   
1307         1308       NaN       3   
1308         1309       NaN       3   

                                                   Name     Sex   Age  SibSp  \
0                               Braund, Mr. Owen Harris    male  22.0      1   
1     Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                                Heikkinen, Miss. Laina  female  26.0      0   
3          Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                              Allen, Mr. William Henry    male  35.0      0   
...                                                 ...     ...   ...    ...   
1304                                 Spector, Mr. Woolf    male  29.0      0   
1305                       Oliva y Ocana, Dona. Fermina  female  39.0      0   
1306                       Saether, Mr. Simon Sivertsen    male  38.5      0   
1307                                Ware, Mr. Frederick    male  29.0      0   
1308                           Peter, Master. Michael J    male  29.0      1   

      Parch              Ticket      Fare Cabin Embarked  Sex_female  \
0         0           A/5 21171    7.2500     0        S           0   
1         0            PC 17599   71.2833   C85        C           1   
2         0    STON/O2. 3101282    7.9250     0        S           1   
3         0              113803   53.1000  C123        S           1   
4         0              373450    8.0500     0        S           0   
...     ...                 ...       ...   ...      ...         ...   
1304      0           A.5. 3236    8.0500     0        S           0   
1305      0            PC 17758  108.9000  C105        C           1   
1306      0  SOTON/O.Q. 3101262    7.2500     0        S           0   
1307      0              359309    8.0500     0        S           0   
1308      1                2668   22.3583     0        C           0   

      Sex_male  Embarked_C  Embarked_Q  Embarked_S  
0            1           0           0           1  
1            0           1           0           0  
2            0           0           0           1  
3            0           0           0           1  
4            1           0           0           1  
...        ...         ...         ...         ...  
1304         1           0           0           1  
1305         0           1           0           0  
1306         1           0           0           1  
1307         1           0           0           1  
1308         1           1           0           0  

[1309 rows x 17 columns]

Para as colunas "Ticket" e "Cabin" o trabalho será mais árduo, pois teremos que fazer um pré-processamento antes da codificação, devido a grande quantidade de valores diferenciados nessas colunas, veja:

In [568]:
# Verificando os valores únicos em "Ticket" e "Cabin"
print('Valores únicos em Ticket:', len(dataset['Ticket'].value_counts()))
print('Valores únicos em Cabin:', len(dataset['Cabin'].value_counts()))

Valores únicos em Ticket: 929
Valores únicos em Cabin: 187


#### Vamos proceder com a transformação da coluna Ticket

In [569]:
# Criando uma nova coluna "Ticket_value" e outra "Ticket_number"
dataset['Ticket_value'] = dataset['Ticket'].str.split(' ').str[0]
dataset['Ticket_number'] = dataset['Ticket'].str.split(' ').str[1]

# Copiando valores de "Ticket_value" para "Ticket_number" quando estiver com valor NaN
dataset['Ticket_number'].fillna(dataset['Ticket_value'], inplace=True)

# Transformando valores numéricos em NaN na coluna "Ticket_value"
dataset['Ticket_value'] = np.where(pd.to_numeric(dataset['Ticket_value'], errors='coerce').notnull(), np.nan, dataset['Ticket_value'])

# Exibindo as primeiras linhas do DataFrame com a coluna atualizada
dataset

PassengerId  Survived  Pclass  \
0               1       0.0       3   
1               2       1.0       1   
2               3       1.0       3   
3               4       1.0       1   
4               5       0.0       3   
...           ...       ...     ...   
1304         1305       NaN       3   
1305         1306       NaN       1   
1306         1307       NaN       3   
1307         1308       NaN       3   
1308         1309       NaN       3   

                                                   Name     Sex   Age  SibSp  \
0                               Braund, Mr. Owen Harris    male  22.0      1   
1     Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                                Heikkinen, Miss. Laina  female  26.0      0   
3          Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                              Allen, Mr. William Henry    male  35.0      0   
...                                                 ...     ...   ...    ...   
1304                                 Spector, Mr. Woolf    male  29.0      0   
1305                       Oliva y Ocana, Dona. Fermina  female  39.0      0   
1306                       Saether, Mr. Simon Sivertsen    male  38.5      0   
1307                                Ware, Mr. Frederick    male  29.0      0   
1308                           Peter, Master. Michael J    male  29.0      1   

      Parch              Ticket      Fare Cabin Embarked  Sex_female  \
0         0           A/5 21171    7.2500     0        S           0   
1         0            PC 17599   71.2833   C85        C           1   
2         0    STON/O2. 3101282    7.9250     0        S           1   
3         0              113803   53.1000  C123        S           1   
4         0              373450    8.0500     0        S           0   
...     ...                 ...       ...   ...      ...         ...   
1304      0           A.5. 3236    8.0500     0        S           0   
1305      0            PC 17758  108.9000  C105        C           1   
1306      0  SOTON/O.Q. 3101262    7.2500     0        S           0   
1307      0              359309    8.0500     0        S           0   
1308      1                2668   22.3583     0        C           0   

      Sex_male  Embarked_C  Embarked_Q  Embarked_S Ticket_value Ticket_number  
0            1           0           0           1          A/5         21171  
1            0           1           0           0           PC         17599  
2            0           0           0           1     STON/O2.       3101282  
3            0           0           0           1          NaN        113803  
4            1           0           0           1          NaN        373450  
...        ...         ...         ...         ...          ...           ...  
1304         1           0           0           1         A.5.          3236  
1305         0           1           0           0           PC         17758  
1306         1           0           0           1   SOTON/O.Q.       3101262  
1307         1           0           0           1          NaN        359309  
1308         1           1           0           0          NaN          2668  

[1309 rows x 19 columns]

In [570]:
# Verificando os valores únicos em "Ticket Value"
dataset['Ticket_value'].value_counts()

PC            92
C.A.          46
SOTON/O.Q.    16
W./C.         14
STON/O        14
A/5           12
CA.           12
SC/PARIS      11
CA            10
A/5.          10
F.C.C.         9
SOTON/OQ       8
C              8
STON/O2.       7
S.O.C.         7
S.O./P.P.      7
A/4            6
SC/Paris       5
SC/AH          5
LINE           4
PP             4
SOTON/O2       3
S.C./PARIS     3
F.C.           3
A.5.           3
A./5.          3
A/4.           3
W.E.P.         2
P/PP           2
SC             2
WE/P           2
STON/OQ.       1
AQ/4           1
SC/A.3         1
C.A./SOTON     1
A.             1
LP             1
SC/A4          1
W/C            1
S.W./PP        1
A/S            1
SCO/W          1
SW/PP          1
S.C./A.4.      1
Fa             1
S.O.P.         1
A4.            1
SO/C           1
S.P.           1
AQ/3.          1
Name: Ticket_value, dtype: int64

In [571]:
# Criando um mapeamento de valores únicos para códigos começando de 1
unique_values = dataset['Ticket_value'].unique()
value_to_code = {value: code+1 for code, value in enumerate(unique_values)}

# Aplicando o mapeamento para criar a nova coluna codificada
dataset['Ticket_value_encoded'] = dataset['Ticket_value'].map(value_to_code).fillna(0).astype(int)

# Exibindo as primeiras linhas do DataFrame com a coluna codificada
dataset

PassengerId  Survived  Pclass  \
0               1       0.0       3   
1               2       1.0       1   
2               3       1.0       3   
3               4       1.0       1   
4               5       0.0       3   
...           ...       ...     ...   
1304         1305       NaN       3   
1305         1306       NaN       1   
1306         1307       NaN       3   
1307         1308       NaN       3   
1308         1309       NaN       3   

                                                   Name     Sex   Age  SibSp  \
0                               Braund, Mr. Owen Harris    male  22.0      1   
1     Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                                Heikkinen, Miss. Laina  female  26.0      0   
3          Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                              Allen, Mr. William Henry    male  35.0      0   
...                                                 ...     ...   ...    ...   
1304                                 Spector, Mr. Woolf    male  29.0      0   
1305                       Oliva y Ocana, Dona. Fermina  female  39.0      0   
1306                       Saether, Mr. Simon Sivertsen    male  38.5      0   
1307                                Ware, Mr. Frederick    male  29.0      0   
1308                           Peter, Master. Michael J    male  29.0      1   

      Parch              Ticket      Fare Cabin Embarked  Sex_female  \
0         0           A/5 21171    7.2500     0        S           0   
1         0            PC 17599   71.2833   C85        C           1   
2         0    STON/O2. 3101282    7.9250     0        S           1   
3         0              113803   53.1000  C123        S           1   
4         0              373450    8.0500     0        S           0   
...     ...                 ...       ...   ...      ...         ...   
1304      0           A.5. 3236    8.0500     0        S           0   
1305      0            PC 17758  108.9000  C105        C           1   
1306      0  SOTON/O.Q. 3101262    7.2500     0        S           0   
1307      0              359309    8.0500     0        S           0   
1308      1                2668   22.3583     0        C           0   

      Sex_male  Embarked_C  Embarked_Q  Embarked_S Ticket_value Ticket_number  \
0            1           0           0           1          A/5         21171   
1            0           1           0           0           PC         17599   
2            0           0           0           1     STON/O2.       3101282   
3            0           0           0           1          NaN        113803   
4            1           0           0           1          NaN        373450   
...        ...         ...         ...         ...          ...           ...   
1304         1           0           0           1         A.5.          3236   
1305         0           1           0           0           PC         17758   
1306         1           0           0           1   SOTON/O.Q.       3101262   
1307         1           0           0           1          NaN        359309   
1308         1           1           0           0          NaN          2668   

      Ticket_value_encoded  
0                        1  
1                        2  
2                        3  
3                        4  
4                        4  
...                    ...  
1304                    25  
1305                     2  
1306                    22  
1307                     4  
1308                     4  

[1309 rows x 20 columns]

In [572]:
# Concatenando a coluna "Ticket_value_encoded" com a coluna "Ticket_number" em "Ticket_final"
dataset['Ticket_final'] = dataset['Ticket_number'] + dataset['Ticket_value_encoded'].astype(str)

# Transformando valores de "Ticket_final" para numéricos
dataset['Ticket_final'] = pd.to_numeric(dataset['Ticket_final'], errors='coerce')

# Exibindo as primeiras linhas do DataFrame com a coluna numérica "Ticket_final"
dataset.head(3)

PassengerId  Survived  Pclass  \
0            1       0.0       3   
1            2       1.0       1   
2            3       1.0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Sex_female  Sex_male  \
0      0         A/5 21171   7.2500     0        S           0         1   
1      0          PC 17599  71.2833   C85        C           1         0   
2      0  STON/O2. 3101282   7.9250     0        S           1         0   

   Embarked_C  Embarked_Q  Embarked_S Ticket_value Ticket_number  \
0           0           0           1          A/5         21171   
1           1           0           0           PC         17599   
2           0           0           1     STON/O2.       3101282   

   Ticket_value_encoded  Ticket_final  
0                     1      211711.0  
1                     2      175992.0  
2                     3    31012823.0

#### Agora vamos transformar a coluna "Cabin" em numérica

In [576]:
dataset['Cabin'].value_counts()

0                  1014
C23 C25 C27           6
B57 B59 B63 B66       5
G6                    5
F33                   4
                   ... 
A14                   1
E63                   1
E12                   1
E38                   1
C105                  1
Name: Cabin, Length: 187, dtype: int64